jobapi/routes.go

```diff
@@ -0,0 +1,134 @@
+package jobapi
+
+import (
+	"encoding/json"
+	"fmt"
+	"net/http"
+
+	"github.com/buildkite/agent/v3/agent"
+	"github.com/go-chi/chi/v5"
+	"github.com/go-chi/chi/v5/middleware"
+	"golang.org/x/exp/maps"
+)
+
+var protectedEnv map[string]struct{}
+
+func init() {
+	protectedEnv = make(map[string]struct{}, len(agent.ProtectedEnv))
+	for _, p := range agent.ProtectedEnv {
+		protectedEnv[p] = struct{}{}
+	}
+}
+
+func (s *Server) router() chi.Router {
+	r := chi.NewRouter()
+	r.Use(
+		middleware.Recoverer,
+		// middleware.Logger, // REVIEW: Should we log requests to this API? If so, where should we log them to? The job logs?
+		HeadersMdlw(),
+		AuthMdlw(s.token),
+	)
+
+	r.Route("/api/current-job/v0", func(r chi.Router) {
+		r.Get("/env", s.getEnv())
+		r.Patch("/env", s.patchEnv())
+		r.Delete("/env", s.deleteEnv())
+	})
+
+	return r
+}
+
+func (s *Server) getEnv() http.HandlerFunc {
+	return func(w http.ResponseWriter, _ *http.Request) {
+		normalizedEnv := s.environ.Dump()
+		resp := EnvGetResponse{Env: normalizedEnv}
+		w.WriteHeader(http.StatusOK)
+		json.NewEncoder(w).Encode(resp)
+	}
+}
+
+func (s *Server) patchEnv() http.HandlerFunc {
+	return func(w http.ResponseWriter, r *http.Request) {
+		var req EnvUpdateRequest
+		err := json.NewDecoder(r.Body).Decode(&req)
+		defer r.Body.Close()
+		if err != nil {
+			err = fmt.Errorf("failed to decode request body: %w", err)
+			returnError(w, err, http.StatusBadRequest)
+			return
+		}
+
+		added := make([]string, len(req.Env))
+		updated := make([]string, len(req.Env))
+		protected := checkProtected(maps.Keys(req.Env))
+
+		if len(protected) > 0 {
+			err = fmt.Errorf("the following environment variables are protected, and cannot be modified: % v", protected)
+			returnError(w, err, http.StatusUnprocessableEntity)
+			return
+		}
+
+		for k, v := range req.Env {
```

Comment:
At current, when the json `{"foo": null}` is unmarshalled into a `map[string]string`, it comes out as `map[string]string{"foo": ""}`, which is a different (and valid) thing to do with the patch API, whereas we don't want to allow people to set env to nil using the patch endpoint - they should be doing that via the delete endpoint.

instead, we should update the Env field of the `EnvUpdateRequest` struct to be a `map[string]*string`, and in this function, check that none of the values are nil. if any are, we should fail the entire operation prior to updating any env.